File: first/func.cs
public static (int, int) DiagNumberAndStart(int k)
{
    var n = 1; 
    var count = 1;
    while (count < k)
    {
        n++;
        count += n;
    }
    return (n, count - n + 1);
}
static int J(int k)
{
    var (diagNum, diagStart) = DiagNumberAndStart(k);
    return k - diagStart + 1;
}
static int I(int k)
{
    var (diagNum, diagStart) = DiagNumberAndStart(k);
    return diagNum + 1 - (k - diagStart + 1);
}

File: first/func.pas
function DiagNumberAndStart(k: integer): (integer, integer);
var
  n: integer;
  count: integer;
begin
  n := 1;
  count := 1;
  while count < k do
  begin
    n := n + 1;
    count := count + n;
  end;
  Result := (n, count - n + 1);
end;

function J(k: integer): integer;
var
  diagNum: integer;
  diagStart: integer;
begin
  (diagNum, diagStart) := DiagNumberAndStart(k);
  Result := k - diagStart + 1;
end;

function I(k: integer): integer;
var
  diagNum: integer;
  diagStart: integer;
begin
  (diagNum, diagStart) := DiagNumberAndStart(k);
  Result := diagNum + 1 - (k - diagStart + 1);
end.